File: java/adventofcode/src/main/java/com/github/mmacdonnacha/day01/Day01.java
package com.github.mmacdonnacha.day01;

import com.github.mmacdonnacha.util.Util;
import java.util.List;

public class Day01 {

    public static void main(String[] args) {
        Util util = new Util();
        Day01 day = new Day01();
        String fileName = "day01.txt";

        List<String> input = util.getInputAsList(fileName);
        List<Integer> data = util.convertToInt(input);


        int part1 = day.solvePart1(data);
        System.out.println("Part 1: " + part1);

        int part2 = day.solvePart2(data);
        System.out.println("Part 2: " + part2);
    }


    public int solvePart1(List<Integer> depths){
        if(depths.isEmpty() || depths.size() == 1)
            return 0;

        int count = 0;

        for(int i=0; i<depths.size()-1; i++){
            if(depths.get(i) < depths.get(i+1)){
                count++;
            }
        }

        return count;
    }

    public int solvePart2(List<Integer> depths){
        if(depths.isEmpty() || depths.size() < 4)
            return 0;

        int count = 0;

        for(int i=0; i<depths.size()-3; i++){
            int firstWindow = depths.get(i) + depths.get(i+1) + depths.get(i+2);
            int secondWindow = depths.get(i+1) + depths.get(i+2) + depths.get(i+3);
        
            if(firstWindow < secondWindow){
                count++;
            }
        }

        return count;
    }
}

File: python/day01/day01.py
from pathlib import Path


def get_file_contents() -> list:
    path = Path(__file__).parent / "../resources/day01.txt"
    with path.open() as f:
        lines = f.readlines()

    return lines


def solve_part1(depths: list) -> int:
    if len(depths) <= 1:
        return 0

    count = 0

    for i in range(len(depths)-1):
        if depths[i] < depths[i+1]:
            count = count + 1

    return count


def solve_part2(depths: list) -> int:
    if len(depths) < 4:
        return 0

    count = 0

    for i in range(len(depths)-3):
        first_window = depths[i] + depths[i+1] + depths[i+2]
        second_window = depths[i+1] + depths[i+2] + depths[i+3]

        if first_window < second_window:
            count = count + 1

    return count


if __name__ == '__main__':
    lines = get_file_contents()
    depths = [int(x) for x in lines]

    part1 = solve_part1(depths)
    print(f'Part 1: {part1}')

    part2 = solve_part2(depths)
    print(f'Part 2: {part2}')
